File: config/validation.py
"""Reusable server-side validators (mirrors static/js/validation.js).

Usage:
    errors = validate(form, {
        "full_name": [("required",), ("alpha",)],
        "national_id": [("required",), ("numeric",)],
        "email": [("required",), ("email",)],
        "dob": [("required",), ("date_not_future",)],
    })
    if errors: raise ValidationError(errors)
"""

import re
from datetime import date, datetime

_EMAIL_RE = re.compile(r"^[^@\s]+@[^@\s]+\.[^@\s]+$")
_ALPHA_RE = re.compile(r"^[A-Za-z][A-Za-z .'-]*$")   # letters, spaces, . ' -
_NUMERIC_RE = re.compile(r"^[0-9]+$")
_BEDNUM_RE = re.compile(r"^[A-Za-z]{2,4}-\d{2,3}$")  # e.g. ER-01, ICU-12
_CNIC_RE = re.compile(r"^\d{13}$")                   # exactly 13 digits


class ValidationError(Exception):
    """Carries a list of human-readable field errors."""

    def __init__(self, errors):
        self.errors = errors if isinstance(errors, list) else [errors]
        super().__init__("; ".join(self.errors))


_LABELS = {
    "full_name": "Name", "fname": "First name", "lname": "Last name",
    "national_id": "CNIC", "phone": "Phone", "email": "Email",
    "dob": "Date of birth", "bed_number": "Bed number", "license_no": "License No",
}


def _label(field):
    return _LABELS.get(field, field.replace("_", " ").title())


def _check(rule, value):
    """Return an error fragment for a single rule, or None if valid."""
    v = (value or "").strip()
    name = rule[0]
    if name == "required":
        return "is required" if v == "" else None
    if v == "":
        return None  # other rules pass on empty (use 'required' to forbid empty)
    if name == "alpha":
        return "must contain letters only" if not _ALPHA_RE.match(v) else None
    if name == "numeric":
        return "must contain digits only" if not _NUMERIC_RE.match(v) else None
    if name == "cnic":
        return "must be exactly 13 digits" if not _CNIC_RE.match(v) else None
    if name == "email":
        return "must be a valid email address" if not _EMAIL_RE.match(v) else None
    if name == "bed_number":
        return "must look like 'ER-01'" if not _BEDNUM_RE.match(v) else None
    if name == "min_len":
        return f"must be at least {rule[1]} characters" if len(v) < rule[1] else None
    if name == "max_len":
        return f"must be at most {rule[1]} characters" if len(v) > rule[1] else None
    if name == "positive_int":
        return ("must be a positive whole number"
                if not (_NUMERIC_RE.match(v) and int(v) > 0) else None)
    if name == "non_negative":
        return ("cannot be negative"
                if not (_NUMERIC_RE.match(v) and int(v) >= 0) else None)
    if name == "date_not_future":
        try:
            d = datetime.strptime(v, "%Y-%m-%d").date()
        except ValueError:
            return "must be a valid date (YYYY-MM-DD)"
        return "cannot be in the future" if d > date.today() else None
    if name == "date":
        try:
            datetime.strptime(v, "%Y-%m-%d")
        except ValueError:
            return "must be a valid date (YYYY-MM-DD)"
        return None
    return None


def validate(data, rules):
    """Run `rules` against `data` (a dict-like). Return a list of error strings."""
    errors = []
    for field, field_rules in rules.items():
        value = data.get(field) if hasattr(data, "get") else data[field]
        for rule in field_rules:
            frag = _check(rule, value)
            if frag:
                errors.append(f"{_label(field)} {frag}.")
                break  # one error per field
    return errors

File: static/js/validation.js
/* HEBAS client-side validation — mirrors config/validation.py.
   Opt in by adding `novalidate data-validate` to a <form> and `data-rule="..."`
   to inputs (space-separated: required alpha numeric email bed date-not-future). */
(function () {
  "use strict";

  const RE = {
    alpha: /^[A-Za-z][A-Za-z .'-]*$/,
    numeric: /^[0-9]+$/,
    email: /^[^@\s]+@[^@\s]+\.[^@\s]+$/,
    bed: /^[A-Za-z]{2,4}-\d{2,3}$/,
    cnic: /^\d{13}$/,
  };
  const MSG = {
    required: "This field is required.",
    alpha: "Letters only.",
    name: "Letters only, at least 2 characters.",
    numeric: "Digits only.",
    cnic: "CNIC must be exactly 13 digits.",
    phone: "Digits only, max 11.",
    email: "Enter a valid email.",
    bed: "Format like ER-01.",
    positive_int: "Enter a positive whole number.",
    non_negative: "Cannot be negative.",
    "date-not-future": "Date cannot be in the future.",
  };

  function fieldError(input) {
    const rules = (input.dataset.rule || "").split(/\s+/).filter(Boolean);
    const v = (input.value || "").trim();
    for (const r of rules) {
      if (r === "required" && v === "") return MSG.required;
      if (v === "") continue;
      if (r === "alpha" && !RE.alpha.test(v)) return MSG.alpha;
      if (r === "name" && (!RE.alpha.test(v) || v.length < 2)) return MSG.name;
      if (r === "numeric" && !RE.numeric.test(v)) return MSG.numeric;
      if (r === "cnic" && !RE.cnic.test(v)) return MSG.cnic;
      if (r === "phone" && (!RE.numeric.test(v) || v.length > 11)) return MSG.phone;
      if (r === "email" && !RE.email.test(v)) return MSG.email;
      if (r === "bed" && !RE.bed.test(v)) return MSG.bed;
      if (r === "positive_int" && !(RE.numeric.test(v) && +v > 0)) return MSG.positive_int;
      if (r === "non_negative" && !(RE.numeric.test(v) && +v >= 0)) return MSG.non_negative;
      if (r === "date-not-future" && v && new Date(v) > new Date(new Date().toDateString()))
        return MSG["date-not-future"];
    }
    return null;
  }

  function applyFeedback(input, err) {
    input.classList.toggle("is-invalid", !!err);
    let fb = input.parentNode.querySelector(".invalid-feedback");
    if (err) {
      if (!fb) {
        fb = document.createElement("div");
        fb.className = "invalid-feedback";
        input.parentNode.appendChild(fb);
      }
      fb.textContent = err;
    }
  }

  function validateForm(form) {
    let ok = true;
    form.querySelectorAll("[data-rule]").forEach((input) => {
      const err = fieldError(input);
      applyFeedback(input, err);
      if (err) ok = false;
    });
    return ok;
  }

  document.addEventListener("DOMContentLoaded", function () {
    document.querySelectorAll("form[data-validate]").forEach((form) => {
      form.addEventListener("submit", function (e) {
        if (!validateForm(form)) {
          e.preventDefault();
          e.stopPropagation();
        }
      });
      // Live de-error on input.
      form.querySelectorAll("[data-rule]").forEach((input) => {
        input.addEventListener("input", () => applyFeedback(input, fieldError(input)));
      });
      // Block invalid characters while typing.
      const mask = (sel, fn) => form.querySelectorAll(sel).forEach((i) =>
        i.addEventListener("input", () => { i.value = fn(i.value); }));
      mask('[data-rule~="numeric"]', (s) => s.replace(/[^0-9]/g, ""));
      mask('[data-rule~="positive_int"]', (s) => s.replace(/[^0-9]/g, ""));
      mask('[data-rule~="non_negative"]', (s) => s.replace(/[^0-9]/g, ""));
      mask('[data-rule~="cnic"]', (s) => s.replace(/[^0-9]/g, "").slice(0, 13));
      mask('[data-rule~="phone"]', (s) => s.replace(/[^0-9]/g, "").slice(0, 11));
      mask('[data-rule~="alpha"]', (s) => s.replace(/[^A-Za-z .'-]/g, ""));
      mask('[data-rule~="name"]', (s) => s.replace(/[^A-Za-z .'-]/g, ""));
    });
  });

  window.HEBAS_validateForm = validateForm;
})();
